Implement an `@export_input` annotation for exporting `StringName/String` as an enum of input actions

Dynamic-Pistol opened the pull request:
Based off https://github.com/godotengine/godot-proposals/discussions/7559, this adds a new annotation which allows exporting string based variables as an enum down which contains every input action from the project input map, this also has an option to allow either built-in actions in the enum dropdown.

Docs are also included, but could be better improved on.

*Bugsquad edit: Superseded by https://github.com/godotengine/godot/pull/96611*

Dynamic-Pistol committed:
Added @export_input for StringName

Added @export_input for StringName based off https://github.com/godotengine/godot-proposals/discussions/7559

Dynamic-Pistol committed:
Added support for `String`

Dynamic-Pistol committed:
Fixed formatting

Calinou commented:
The boolean type as a parameter is hard to understand if you don't know what it's supposed to type. Like `@rpc()` parameters, I think a string parameter like `"allow_builtin"` would be preferable here.

dalexeev commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -669,6 +669,15 @@
 		</annotation>
+		<annotation name="@export_input">
+			<description>
+				Exports a String or StringName as an enum of the current project's InputMap.
```
```suggestion
				Exports a [StringName], [String], [Array][lb][StringName][rb], [Array][lb][String][rb], or [PackedStringArray] property as an enum of the current project's [InputMap] actions.
```

dalexeev commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -669,6 +669,15 @@
+			<description>
+				Exports a String or StringName as an enum of the current project's InputMap.
+				[codeblock]
+				@export_input var input_action : StringName
+				@export_input(true) var builtin_action: String
```
```suggestion
				@export_input var input_action: StringName
				@export_input var input_actions: Array[StringName]
				@export_input(true) var builtin_action: String
```

dalexeev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -122,6 +122,7 @@ GDScriptParser::GDScriptParser() {
 		register_annotation(MethodInfo("@export_flags_avoidance"), AnnotationInfo::VARIABLE, &GDScriptParser::export_annotations<PROPERTY_HINT_LAYERS_AVOIDANCE, Variant::INT>);
 		register_annotation(MethodInfo("@export_storage"), AnnotationInfo::VARIABLE, &GDScriptParser::export_storage_annotation);
 		register_annotation(MethodInfo("@export_custom", PropertyInfo(Variant::INT, "hint", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_CLASS_IS_ENUM, "PropertyHint"), PropertyInfo(Variant::STRING, "hint_string"), PropertyInfo(Variant::INT, "usage", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_CLASS_IS_BITFIELD, "PropertyUsageFlags")), AnnotationInfo::VARIABLE, &GDScriptParser::export_custom_annotation, varray(PR ...
+		register_annotation(MethodInfo("@export_input", PropertyInfo(Variant::BOOL, "built_in")), AnnotationInfo::VARIABLE, &GDScriptParser::export_annotations<PROPERTY_HINT_INPUT_NAME, Variant::STRING>, varray(false));
```
```suggestion
		register_annotation(MethodInfo("@export_input", PropertyInfo(Variant::BOOL, "built_in")), AnnotationInfo::VARIABLE, &GDScriptParser::export_annotations<PROPERTY_HINT_INPUT_NAME, Variant::STRING_NAME>, varray(false));
```

dalexeev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -4416,6 +4417,12 @@ bool GDScriptParser::export_annotations(const AnnotationNode *p_annotation, Node
 			push_error(_get_annotation_error_string(p_annotation->name, expected_types, variable->get_datatype()), p_annotation);
 			return false;
 		}
+	} else if (p_annotation->name == SNAME("@export_input")) {
```
```suggestion
	} else if (p_annotation->name == SNAME("@export_input")) {
		use_default_variable_type_check = false;

```

dalexeev commented on modules/gdscript/gdscript_parser.cpp:
This annotation handles arguments differently than other export annotations. You need to add a check above.

dalexeev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -4416,6 +4417,12 @@ bool GDScriptParser::export_annotations(const AnnotationNode *p_annotation, Node
 			return false;
 		}
+	} else if (p_annotation->name == SNAME("@export_input")) {
+		if (!(variable->export_info.type == Variant::STRING_NAME || variable->export_info.type == Variant::STRING)) {
```
I haven't tested, but it should be something like:

```suggestion
		if (!export_type.is_variant() && !(export_type.builtin_type == Variant::STRING_NAME || export_type.builtin_type == Variant::STRING)) {
```

AThousandShips replied:
```suggestion
		if (!export_type.is_variant() && export_type.builtin_type != Variant::STRING_NAME && export_type.builtin_type != Variant::STRING) {
```
If the above is valid I'd say for clarity

dalexeev reviewed (commented):
I'm not sure this is really necessary. Linked discussion is quite short. Should it be called `@export_input_action` instead? Perhaps this should be turned into a full proposal first?

Dynamic-Pistol commented on editor/editor_properties.cpp:
```diff
@@ -3651,6 +3652,21 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+					if ((p_hint_text == "false" || p_hint_text.is_empty()) && InputMap::get_singleton()->get_builtins().has(action_name)) {
+						continue;
+					}
+					if (((String)action_name).begins_with("spatial_editor/")) {
```
The actions are related to the editor i assume,they typically aren't meant to be used in gameplay i assume, should i keep them?

Dynamic-Pistol commented on editor/editor_properties.cpp:
```diff
@@ -3651,6 +3652,21 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+			} else if (p_hint == PROPERTY_HINT_INPUT_NAME) {
+				EditorPropertyTextEnum *editor = memnew(EditorPropertyTextEnum);
+				Vector<String> options;
+				InputMap::get_singleton()->load_from_project_settings();
```
You mean calling `InputMap::get_singleton()->load_from_project_settings();` ? it's the only way to get custom input actions here, unless there is another function that is better for this case.

bruvzg replied:
There's no reason to load it, and this will overwrite editor input map with project (and will ruin editor setting if you have default actions override in the project), and show editor actions in the list, so this is definitely the wrong way to do it. Instead, you can load the list directly, see `load_from_project_settings` implementation, something like:

```c++
	EditorPropertyTextEnum *editor = memnew(EditorPropertyTextEnum);
	Vector<String> options;
	List<PropertyInfo> pinfo;
	ProjectSettings::get_singleton()->get_property_list(&pinfo);

	for (const PropertyInfo &pi : pinfo) {
		if (!pi.name.begins_with("input/")) {
			continue;
		}

		String action_name = pi.name.substr(pi.name.find("/") + 1, pi.name.length());
		if ((p_hint_text != "allow_builtin") && InputMap::get_singleton()->get_builtins().has(action_name)) {
			continue;
		}
		options.append(action_name);
	}
	editor->setup(options, false, true);
	return editor;
```

Dynamic-Pistol commented on editor/editor_properties.cpp:
```diff
@@ -3651,6 +3652,21 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
 				}
 				return editor;
+			} else if (p_hint == PROPERTY_HINT_INPUT_NAME) {
+				EditorPropertyTextEnum *editor = memnew(EditorPropertyTextEnum);
```
Sorry, but i don't get what you mean, do you mean i place it as the first hint or what?

Dynamic-Pistol committed:
Fixed some issues and added dropdown for other input types.

Improved docs, fixed some issues with Property hints, improved the code in general and added support for `TouchScreenButton` and `InputEventAction`, currently only missing 2 requirements before acceptance

AThousandShips commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -577,6 +577,19 @@
+			<description>
+				Exports a [StringName], [String], [Array][lb][StringName][rb], [Array][lb][String][rb], or [PackedStringArray] property as an enum of the current project's [InputMap] actions.
+				If has the hint [code]"allow_builtin"[/code], then it also shows the built-in actions.
+				[codeBlock]
```
```suggestion
				[CodeBlock]
```

AThousandShips commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -577,6 +577,19 @@
+				@export_input var input_action: StringName
+				@export_input var input_actions: Array[StringName]
+				@export_input(true) var builtin_action: String
+				[/codeBlock]
```
```suggestion
				[/codeblock]
```

bruvzg commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2932,7 +2932,9 @@
 		</constant>
-		<constant name="PROPERTY_HINT_MAX" value="38" enum="PropertyHint">
+		<constant name="PROPERTY_HINT_INPUT_NAME" value="38" enum="PropertyHint">
+		</constant>
```
Missing description.

bruvzg commented on editor/editor_properties.cpp:
```diff
@@ -3778,6 +3794,21 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+			} else if (p_hint == PROPERTY_HINT_INPUT_NAME) {
+				EditorPropertyTextEnum *editor = memnew(EditorPropertyTextEnum);
+				Vector<String> options;
+				InputMap::get_singleton()->load_from_project_settings();
```
Same here.

Dynamic-Pistol committed:
Fixed issues and added extra docs

Dynamic-Pistol commented:
@bruvzg i added the changes you requested, anything more?

bruvzg commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -577,6 +577,19 @@
+				Exports a [StringName], [String], [Array][lb][StringName][rb], [Array][lb][String][rb], or [PackedStringArray] property as an enum of the current project's [InputMap] actions.
+				If has the hint [code]"allow_builtin"[/code], then it also shows the built-in actions.
+				[codeblock]
+				@export_input var input_action: StringName
+				@export_input var input_actions: Array[StringName]
+				@export_input("allow_builtin") var builtin_action: String
```
```suggestion
				@export_input_action var input_action: StringName
				@export_input_action var input_actions: Array[StringName]
				@export_input_action("allow_builtin") var builtin_action: String
```

bruvzg commented:
Please squash your commits, see [pull request workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase).

Mickeon commented:
I let this simmer for a bit. I am not against this export annotation, but I'm not fully convinced on it, either.
- First of all, I do think having a full-on proposal may help freshen up ideas more.
    - As already noted in the discussion, it's already _kinda_ possible to replicate what this annotation would do, so it's worth questioning if this is worth being in core.
- The name `@export_input` is misleading. It sounds like it's a special way to export **InputEvent** resources. It should probably something like `@export_action`, `@export_input_action`, `@export_action_name`, etc.
- This annotation can be useful, but scales badly in a large project. A lot of the times you do not want absolutely every action to choose from. A way to filter out only a specific subset of them (e.g. by RegEx) _could_ be interesting.
    - The way enum options are currently displayed in the Inspector does not scale well. There's also nothing like a search bar.
- I'm wondering if a more broad annotation type could be introduced to accomplish _this_ and potentially more. Something more expandable and not "very specific" as is in this PR.

Dynamic-Pistol commented:
@Mickeon 

1. @dalexeev Also said this, I am not sure whether to make a new proposal or continue in the existing one i linked, also as for "Why should be in core?" is because it is quite tedious to do and repetitive and requires `@tool`.
2. Already fixed, title is just outdated.
3. Was thinking of using [EditorNode](https://github.com/godotengine/godot-proposals/discussions/7559#discussioncomment-7768917) like someone said in the proposal I linked, but I don't there is an option for custom options, I/Someone else could add this in the future if a better way exists, but for now this is the best we can have I think, also i ensured for a way to exclude built-in actions to make it clean, also this could be just me, but I don't think i have ever seen a project use more than 15 input actions.
4. I don't know to be honest, maybe `@export("input_action:allow_builtin;")` but not sure, would require a proposal.

Some bonus points:

- I first was going to add a way to set `String/StringName`'s value to input action, looked at proposals and found the linked post, thought it could be easy to add (which is kinda true?) and decided to give it a shot.
- This is my first contribution to godot.
- This feature was mainly looked for because I wanted any easy to set names to input actions easily.

AThousandShips commented on editor/editor_properties.cpp:
```diff
@@ -3626,6 +3627,25 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+						continue;
+					}
+
+					String action_name = pi.name.substr(pi.name.find("/") + 1, pi.name.length());
```
```suggestion
					String action_name = pi.name.substr(pi.name.find("/") + 1);
```
Unnecessary

AThousandShips commented on editor/editor_properties.cpp:
```diff
@@ -3778,6 +3798,25 @@ EditorProperty *EditorInspectorDefaultPlugin::get_editor_for_property(Object *p_
+						continue;
+					}
+
+					String action_name = pi.name.substr(pi.name.find("/") + 1, pi.name.length());
```
```suggestion
					String action_name = pi.name.substr(pi.name.find("/") + 1);
```

Dynamic-Pistol committed:
Update editor/editor_properties.cpp

Co-authored-by: A Thousand Ships (she/her) <96648715+AThousandShips@users.noreply.github.com>

Dynamic-Pistol committed:
Update editor/editor_properties.cpp

Co-authored-by: A Thousand Ships (she/her) <96648715+AThousandShips@users.noreply.github.com>

Calinou reviewed (commented):
Tested locally, it mostly works as expected. This is a good start :slightly_smiling_face: 

Some feedback:

- Built-in actions should be listed *last* when `allow_builtin` is enabled. Otherwise, you have to scroll down the list a lot to find non-built-in actions. I think most of the time, you'll still want to select your own actions as opposed to built-in actions when built-in actions are allowed.
- Should `allow_builtin` be renamed to `show_builtin`? If `allow_builtin` is not used, you can still manually enter the name of a built-in action and the inspector won't prevent you from doing so. I think that behavior is fine, but we should be careful about semantics here.
- When you add/remove/rename an action, the inspector isn't updated automatically. This means it can display outdated state, with actions that don't exist anymore (and newly created actions won't show up). It'll only update when you select another node, or collapse then expand the subresource that contains the input action.

Mickeon commented:
> Should allow_builtin be renamed to show_builtin?

It should be `show_builtin` or maybe `include_builtin`, because "allow" is from the perspective of what the code does internally, but in practice it's about showing them.

> When you add/remove/rename an action, the inspector isn't updated automatically.

Many, many such cases. If this were to be fixed, so should situations where the script is modified but the open Inspector is not updated, and there's no catch-all solution unfortunately.

Dynamic-Pistol commented:
> It should be show_builtin or maybe include_builtin, because "allow" is from the perspective of what the code does internally, but in practice it's about showing them.

I can do that quickly.

> Many, many such cases. If this were to be fixed, so should situations where the script is modified but the open Inspector is not updated, and there's no catch-all solution unfortunately.

I am not sure if i can fix that.

> Built-in actions should be listed last when allow_builtin is enabled. Otherwise, you have to scroll down the list a lot to find non-built-in actions. I think most of the time, you'll still want to select your own actions as opposed to built-in actions when built-in actions are allowed.

I can try doing that.

adamscott commented:
Discussed during a GDScript meeting. We concluded that this PR is super interesting, but we aren't sure that adding a new annotation is the way to go. We should instead take your `PROPERTY_HINT_INPUT_NAME` and expose it for the new `@export_custom` annotation. That way, users can create their own field with input_name values, but without having to create a new annotation just for this.

Dynamic-Pistol commented:
@adamscott 
So is anything needed from me?

Dynamic-Pistol committed:
Revert "Update editor/editor_properties.cpp"

This reverts commit 548e4f0027212ef34bcbd3a3942b9aec77908df9.

Dynamic-Pistol committed:
Revert "Update editor/editor_properties.cpp"

This reverts commit 103602ab9d09af125d8ecf7542af6ca8abdbecc9.

Dynamic-Pistol committed:
Reapply "Update editor/editor_properties.cpp"

This reverts commit 7c26da98003f7e99c4d0a7202a5a56158731431b.

Dynamic-Pistol committed:
Reapply "Update editor/editor_properties.cpp"

This reverts commit 80b392bafb1d60b84c1b082dfba2e03aceac0a04.

Dynamic-Pistol committed:
Removed old annotation docs + Fixed some minor issues with hint docs

Dynamic-Pistol committed:
used --doctool

Dynamic-Pistol committed:
Other stuff i forgot

Dynamic-Pistol committed:
Fix misspell.

Dynamic-Pistol committed:
Fixed tests

Dynamic-Pistol committed:
Revert "used --doctool"

This reverts commit ac110c8066eebee88d8a8e14925a76465c9ac339.

Dynamic-Pistol committed:
Revert "Other stuff i forgot"

This reverts commit 144dd739808a19e4392b7dc53456fc2cce4095d0.

Dynamic-Pistol committed:
Fixed issue

Dynamic-Pistol committed:
Revert "Fixed tests"

This reverts commit 8f782612bb178edfd52f0623f4e2b8acaa81cb1b.

Dynamic-Pistol committed:
Fixed issue again

adamscott commented:
> @adamscott So is anything needed from me?

If you could create a new PR with only your code that creates `PROPERTY_HINT_INPUT_NAME`, it could solve the issue.
